File: model_description/scripts/landmark.py
#!/usr/bin/env python

import rospy, tf, rospkg, random
from gazebo_msgs.srv import DeleteModel, SpawnModel, GetModelState
from geometry_msgs.msg import Quaternion, Pose, Point

class CubeSpawner():

    def __init__(self):
        self.rospack = rospkg.RosPack()
        self.path = self.rospack.get_path('model_description') + "/urdf/"
        self.red_cube = self.path + "red_cube.urdf"
        self.spawned_cubes = []
        self.min_spacing = 1.0
        self.cube_positions = []

        self.sm = rospy.ServiceProxy("/gazebo/spawn_urdf_model", SpawnModel)
        self.dm = rospy.ServiceProxy("/gazebo/delete_model", DeleteModel)

    def random_position(self):
        while True:
            x = random.uniform(-5, 5)
            y = random.uniform(-5, 5)  
            if all(self.min_distance(x, y, cube) for cube in self.spawned_cubes) and not (-0.2 <= x <= 0.2 and -0.2 <= y <= 0.2):
                return x, y


    def min_distance(self, x, y, cube):
        for c in self.spawned_cubes:
            pose = self.get_pose(c)
            dx = x - pose.position.x
            dy = y - pose.position.y
            distance = (dx ** 2 + dy ** 2) ** 0.5
            if distance < self.min_spacing:
                return False
        return True

    def get_pose(self, model_name):
        pose = Pose()
        try:
            pose = rospy.ServiceProxy("/gazebo/get_model_state", GetModelState)(model_name, "world").pose
        except rospy.ServiceException as e:
            rospy.logwarn("Service call failed: %s" % e)
        return pose

    def spawn_model(self):
        x, y = self.random_position()
        with open(self.red_cube, "r") as f:
            cube_urdf = f.read()
        quat = tf.transformations.quaternion_from_euler(0, 0, 1.57)
        orient = Quaternion(quat[0], quat[1], quat[2], quat[3])
        pose = Pose(Point(x=x, y=y, z=0.75), orient)
        model_name = "red_cube_" + str(len(self.spawned_cubes))
        self.sm(model_name, cube_urdf, "", pose, "world")
        self.spawned_cubes.append(model_name)
        self.cube_positions.append(pose)

    def delete_all_models(self):
        for model in self.spawned_cubes:
            self.dm(model)

    def shutdown_hook(self):
        self.delete_all_models()
        print("Shutting down")

if __name__ == "__main__":
    print("Waiting for gazebo services...")
    rospy.init_node("landmark_spawner")
    rospy.wait_for_service("/gazebo/delete_model")
    rospy.wait_for_service("/gazebo/spawn_urdf_model")

    cs = CubeSpawner()
    rospy.on_shutdown(cs.shutdown_hook)
    rate = rospy.Rate(15)

    # Spawn multiple instances of the red cube
    for _ in range(10):
        cs.spawn_model()
        rate.sleep()

    rospy.spin()

File: model_description/scripts/landmark.cpp
#include <ros/ros.h>
#include <ros/package.h>
#include <gazebo_msgs/DeleteModel.h>
#include <gazebo_msgs/SpawnModel.h>
#include <gazebo_msgs/GetModelState.h>
#include <geometry_msgs/Quaternion.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/Point.h>
#include <tf/tf.h>
#include <string>
#include <vector>
#include <random>
#include <fstream>

class CubeSpawner
{
public:
    CubeSpawner(ros::NodeHandle &nh) : nh_(nh)
    {
        rospack_ = ros::package::getPath("model_description");
        path_ = rospack_ + "/urdf/";
        red_cube_ = path_ + "red_cube.urdf";
        min_spacing_ = 1.0;
    }

    void randomPosition(double &x, double &y)
    {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<double> x_dist(-5.0, 5.0);
        std::uniform_real_distribution<double> y_dist(-5.0, 5.0);

        while (true)
        {
            x = x_dist(gen);
            y = y_dist(gen);
            if (minDistance(x, y))
            {
                if (!(x >= -1.0 && x <= 1.0 && y >= -1.0 && y <= 1.0))
                {
                    return;
                }
            }
        }
    }

    bool minDistance(double x, double y)
    {
        for (const std::string &cube : spawned_cubes_)
        {
            geometry_msgs::Pose pose = getPose(cube);
            double dx = x - pose.position.x;
            double dy = y - pose.position.y;
            double distance = std::sqrt(dx * dx + dy * dy);
            if (distance < min_spacing_)
            {
                return false;
            }
        }
        return true;
    }

    geometry_msgs::Pose getPose(const std::string &model_name)
    {
        geometry_msgs::Pose pose;
        gazebo_msgs::GetModelState srv;
        srv.request.model_name = model_name;
        srv.request.relative_entity_name = "world";

        if (get_model_state_client_.call(srv))
        {
            pose = srv.response.pose;
        }
        else
        {
            ROS_WARN("Service call failed");
        }
        return pose;
    }

    void spawnModel()
    {
        double x, y;
        randomPosition(x, y);
        std::string model_name = "red_cube_" + std::to_string(spawned_cubes_.size());
        std::ifstream file(red_cube_);
        std::string cube_urdf((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());

        tf::Quaternion quat;
        quat.setEuler(0, 0, 1.57);
        geometry_msgs::Quaternion orient;
        orient.x = quat.getX();
        orient.y = quat.getY();
        orient.z = quat.getZ();
        orient.w = quat.getW();

        geometry_msgs::Pose pose;
        pose.position.x = x;
        pose.position.y = y;
        pose.position.z = 0.75;
        pose.orientation = orient;

        gazebo_msgs::SpawnModel srv;
        srv.request.model_name = model_name;
        srv.request.model_xml = cube_urdf;
        srv.request.initial_pose = pose;
        srv.request.reference_frame = "world";

        if (spawn_model_client_.call(srv))
        {
            spawned_cubes_.push_back(model_name);
            cube_positions_.push_back(pose);
        }
    }

    void deleteAllModels()
    {
        for (const std::string &model : spawned_cubes_)
        {
            gazebo_msgs::DeleteModel srv;
            srv.request.model_name = model;
            delete_model_client_.call(srv);
        }
    }

    void shutdownHook()
    {
        deleteAllModels();
        ROS_INFO("Shutting down");
    }

    void run()
    {
        ROS_INFO("Waiting for gazebo services...");
        ros::service::waitForService("/gazebo/delete_model");
        ros::service::waitForService("/gazebo/spawn_urdf_model");

        ros::Rate rate(15);

        for (int i = 0; i < 10; ++i)
        {
            spawnModel();
            rate.sleep();
        }

        ros::spin();
    }

private:
    ros::NodeHandle nh_;
    std::string rospack_;
    std::string path_;
    std::string red_cube_;
    double min_spacing_;
    std::vector<std::string> spawned_cubes_;
    std::vector<geometry_msgs::Pose> cube_positions_;
    ros::ServiceClient spawn_model_client_ = nh_.serviceClient<gazebo_msgs::SpawnModel>("/gazebo/spawn_urdf_model");
    ros::ServiceClient delete_model_client_ = nh_.serviceClient<gazebo_msgs::DeleteModel>("/gazebo/delete_model");
    ros::ServiceClient get_model_state_client_ = nh_.serviceClient<gazebo_msgs::GetModelState>("/gazebo/get_model_state");
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "landmark_spawner");
    ros::NodeHandle nh;
    CubeSpawner cs(nh);
    // ros::shutdown();

    cs.run();
    return 0;
}
